### DISCUSSION TITLE
How to do `save_file_dialog` or Share a file on Android?

### DISCUSSION THREAD

**Discussion by vzool** [beeware/toga#2719] [Q&A]
Here is the snippet code that works and opens a dialog, the file is created but it is zero bytes:
```python
    from android.content import Intent
    from android.net import Uri
    from java.io import File

    file_path = os.path.join(self.paths.data, 'zakat.pickle')
    file = File(str(file_path))
    print('file.length:', file.length())

    intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
    intent.setType("*/*")
    intent.addCategory(Intent.CATEGORY_OPENABLE)
    intent.putExtra(Intent.EXTRA_TITLE, 'zakat.pickle')
    intent.putExtra(Intent.EXTRA_STREAM, Uri.fromFile(file));

    self._impl.start_activity(Intent.createChooser(intent, 'Save database'))
```

### Open save Dialog
<img src="https://github.com/user-attachments/assets/84c9b616-b975-4207-8a80-c6e0abe9f8b3" width="400">

### After clicking `Save`, file `zakat.pickle (1)` created but it is zero bytes
<img src="https://github.com/user-attachments/assets/fe305e39-dd1e-411d-8e3b-47cb44bb054f" width="400">

Any suggestions?

Thanks

**Comment by freakboy3742** [comment]
I have no particular experience with the Android File dialog APIs, but my immediate impression is: if the file exists, but there's no content in the file... where is the code that is writing to the file? Based on the snippet you've provided, I can see you've created a Java  `File` object, but you're not invoking anything that would *write* to that file.

**Reply by mhsmith** [reply]
I think he's attempting to attach an existing file to the intent, but that's not the correct way to use ACTION_CREATE_DOCUMENT. You're supposed to wait for the intent result, which will contain a URI that you can write to.

See https://github.com/beeware/toga/pull/1158#issuecomment-2254564657 for a link to a Java example, and https://github.com/beeware/toga/pull/1158#issuecomment-1446196973 for how to wait for an intent result.
